### PR TITLE
Updated functionality for `add_rfree` and `copy_rfree`

### PR DESCRIPTION
**Pull request by dennisbrookner**
Making changes to `rs.utils.add_rfree` as per #169 and changes to `rs.utils.copy_rfree` as per #159 

## `add_rfree` takes `seed`

The `seed` is passed through to `np.random.default_rng(seed)`, such that running `add_rfree` again with the same `seed` results in the same flags. `seed` defaults to `None`, and conveniently `np.random.default_rng(None)` is a valid call.

A test was added to `test_add_rfree()` to confirm that flags are the same when a seed is used, and are different when a seed is not used.

## `copy_rfree` handles column names flexibly

The recent addition of the `ccp4_convention` parameter to `add_rfree` means that there are two possible names for an r-free column, `"R-free-flags"` (phenix convention) or `"FreeR_flag"` (ccp4 convention). `copy_rfree` will now by default search the provided `dataset_with_rfree` for a column named `"R-free-flags"`, then `"FreeR_flag"`, and throw a `ValueError` if neither is found.

The user can override this behavior via the `custom_rfree_key` argument if the r-free flag column uses any other naming convention.

A function `test_copy_rfree()` was added to test successfully copying in all combinations of column naming. A function `test_copy_rfree_errors()` was added to verify that `ValueError`s are thrown when appropriate.

### Old `unittest` code
There is an existing test for `copy_rfree` written with `unittest`. I believe that all key functionalities are tested in my new `pytest` code, making the `unittest` code redundant, but I left it in for now and labelled it in its docstring that it is a "legacy" version. Let me know if there's anything in the `unittest` code that I should additionally migrate, and/or if the `unittest` code should me removed altogether.

### CODE REVIEW

File: tests/utils/test_rfree.py [hunk 1]
```diff
@@ -10,13 +10,18 @@
 @pytest.mark.parametrize("fraction", [0.05, 0.10, 0.15])
 @pytest.mark.parametrize("ccp4_convention", [False, True])
 @pytest.mark.parametrize("inplace", [False, True])
-def test_add_rfree(data_fmodel, fraction, ccp4_convention, inplace):
+@pytest.mark.parametrize("seed", [None, 2022])
+def test_add_rfree(data_fmodel, fraction, ccp4_convention, inplace, seed):
     """
-    Test rs.utils.add_rfee
+    Test rs.utils.add_rfree
```
Comment (JBGreisman) [inline tests/utils/test_rfree.py:16]: nice catch.

File: reciprocalspaceship/utils/rfree.py [hunk 2]
```diff
@@ -46,7 +52,7 @@ def add_rfree(dataset, fraction=0.05, ccp4_convention=False, inplace=False):
     return dataset
 
 
-def copy_rfree(dataset, dataset_with_rfree, inplace=False):
+def copy_rfree(dataset, dataset_with_rfree, inplace=False, custom_rfree_key=None):
```
Comment (JBGreisman) [inline reciprocalspaceship/utils/rfree.py:55]: I think this the parameter should be renamed to `rfree_key` -- I think the "custom" is unnecessary here.

**Review by JBGreisman — CHANGES_REQUESTED** [review]
Thanks for taking care of these -- your changes look good to me. My only updates:
1. Please change `custom_rfree_key` --> `rfree_key`. I would rather keep the function arguments as short as possible, and I don't think the "custom" adds clarification. 
2. Please delete the now obsolete `unittest` code -- your `pytest` versions look good to me.

File: reciprocalspaceship/utils/rfree.py [hunk 3]
```diff
@@ -66,8 +76,22 @@ def copy_rfree(dataset, dataset_with_rfree, inplace=False):
...
+            raise ValueError(
+                f"""Supplied dataset_with_rfree contains no column {rfree_key}"""
+            )
+    elif "R-free-flags" in dataset_with_rfree.columns:
+        rfree_key = "R-free-flags"
+    elif "FreeR_flag" in dataset_with_rfree.columns:
+        rfree_key = "FreeR_flag"
+    else:
+        raise ValueError(
+            """Failed to automatically find r-free flags in dataset_with_rfree. Please supply a custom_rfree_key"""
```
Comment (JBGreisman) [inline reciprocalspaceship/utils/rfree.py:90]: sorry small thing -- do you mind changing `custom_rfree_key` here too in the ValueError?

Comment (dennisbrookner) [reply reciprocalspaceship/utils/rfree.py:90]: Ah missed that, my bad

File: tests/utils/test_rfree.py [hunk 4]
```diff
@@ -38,34 +43,72 @@ def test_add_rfree(data_fmodel, fraction, ccp4_convention, inplace):
...
+def test_copy_rfree_errors(data_fmodel):
+    """
+    Test expected ValueErrors for rs.utils.copy_rfree
+    """
+    with pytest.raises(ValueError):
+        rs.utils.copy_rfree(data_fmodel, data_fmodel)
 
-        return
+    with pytest.raises(ValueError):
+        rs.utils.copy_rfree(data_fmodel, data_fmodel, rfree_key="missing key")
```
Comment (JBGreisman) [inline tests/utils/test_rfree.py:114]: I didn't catch this before, but this can be simplified to:
```python
@pytest.mark.parametrize("rfree_key", [None, "missing key"])
def test_copy_rfree_errors(data_fmodel, rfree_key):
    """
    Test expected ValueErrors for rs.utils.copy_rfree()
    """
    with pytest.raises(ValueError):
        rs.utils.copy_rfree(data_fmodel, data_fmodel, rfree_key=rfree_key)
 ```
 I think the above is true... let me know if I'm mistaken

Comment (dennisbrookner) [reply tests/utils/test_rfree.py:114]: Oo good call, I think you're right. Didn't occur to me because it's a different `ValueError`, but yeah this should work.

Comment (dennisbrookner) [reply tests/utils/test_rfree.py:114]: I'll just clarify in the docstring what the two ValueErrors are, because I think the condensed version is a little harder to parse

Comment (dennisbrookner) [reply tests/utils/test_rfree.py]: In theory a slightly more rigorous test would be to confirm that the "missing key" case throws an error even when "R-free-flag" or "FreeR_flags" is present, but I think that's probably overkill.

Comment (JBGreisman) [reply tests/utils/test_rfree.py]: yep, I think your comment there is good. 

I think the test you just mentioned is a pretty good one because it tests the intended implementation (that `rfree_key` supercedes any other default column names). If we change the order of the `if` clauses in the future, that could change, so I think those sorts of tests are useful for making sure as many corner cases are covered.

File: tests/utils/test_rfree.py [hunk 5]
```diff
@@ -109,7 +109,7 @@ def test_copy_rfree_errors(data_fmodel):
         rs.utils.copy_rfree(data_fmodel, data_fmodel)
 
     # Raise ValueError because "missing key" is missing,
-    # even though "R-free-flags" is present
+    # even though "R-free-flags" exists
```
Comment (dennisbrookner) [inline tests/utils/test_rfree.py:112]: Had to make a change to get out of a loop of git madness, figured this was benign enough

**Review by JBGreisman — APPROVED** [review]
Looks good -- thanks for making the fixes!
